datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1225,12 +1228,33 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
         _adapter.updateAllAssignments(newAssignmentsByInstance);
       }
 
+      Set<String> failedStreams = Collections.emptySet();
+      // Poll the zookeeper to ensure that hosts claimed assignment tokens for stopping streams
+      if (_config.getEnableAssignmentTokens() &&
+          !PollUtils.poll(() -> _adapter.getNumUnclaimedTokensForDatastreams(stoppingDatastreamGroups) == 0,
+            STOP_PROPAGATION_RETRY_MS, _config.getStopPropagationTimeout())) {
+        Map<String, List<AssignmentToken>> unclaimedTokens =
+            _adapter.getUnclaimedAssignmentTokensForDatastreams(stoppingDatastreamGroups);
+        failedStreams = unclaimedTokens.keySet();
+        Set<String> hosts = unclaimedTokens.values().stream().flatMap(List::stream).
+            map(AssignmentToken::getIssuedFor).collect(Collectors.toSet());
+        _log.error("Stop failed to propagate within {}ms for streams: {}. The following hosts failed to claim their token(s): {}",
+            _config.getStopPropagationTimeout(), failedStreams, hosts);
+        revokeUnclaimedAssignmentTokens(unclaimedTokens);
+        _metrics.updateMeter(CoordinatorMetrics.Meter.NUM_FAILED_STOPS, failedStreams.size());
+      }
+
+      // TODO Explore introduction of UNKNOWN/WARN state for streams that failed to stop
```

Comment:
Should we keep the streams in the 'STOPPING" state if the tokens cannot be claimed?

Should we fall back to the current way of stopping streams if any of this ZK business goes wrong in worst-case scenarios? And since we have a 60 seconds timeout anyway, should we mark the stream as Stopped (as we currently do)?

Reply:
On a second thought, I don't think we should do that.

Reply:
I was having the same debate. I think it's too risky to change that behavior. I'm planning to get back to this and introduce a new state in the state machine after testing the rest of it (at least the metrics etc need to work really well).